File: 01_process_thread/php/app/Http/Controllers/HeavyController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\JsonResponse;

class HeavyController extends Controller
{
    private int $heavyCalcN;

    public function __construct()
    {
        $envValue = getenv('HEAVY_CALC_N');
        $this->heavyCalcN = $envValue !== false ? (int) $envValue : 40;
    }

    private function fibonacci(int $n): int
    {
        if ($n <= 1) {
            return $n;
        }
        return $this->fibonacci($n - 1) + $this->fibonacci($n - 2);
    }

    public function health(): JsonResponse
    {
        return response()->json([
            'status' => 'ok',
            'language' => 'php',
        ]);
    }

    public function heavy(): JsonResponse
    {
        $startedAt = gmdate('Y-m-d\TH:i:s.v\Z');
        $startMs = hrtime(true);

        $this->fibonacci($this->heavyCalcN);

        $endMs = hrtime(true);
        $finishedAt = gmdate('Y-m-d\TH:i:s.v\Z');

        $durationMs = (int) (($endMs - $startMs) / 1_000_000);

        return response()->json([
            'language' => 'php',
            'threadId' => 'fpm-worker-' . getmypid(),
            'startedAt' => $startedAt,
            'finishedAt' => $finishedAt,
            'durationMs' => $durationMs,
        ]);
    }
}

File: 01_process_thread/ruby/app/controllers/heavy_controller.rb
class HeavyController < ApplicationController
  HEAVY_CALC_N = ENV.fetch("HEAVY_CALC_N", "40").to_i

  def health
    render json: { status: "ok", language: "ruby" }
  end

  def heavy
    n = params[:n].present? ? params[:n].to_i : HEAVY_CALC_N
    started_at = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%S.%3NZ")
    start_ns = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

    fibonacci(n)

    end_ns = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
    finished_at = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%S.%3NZ")

    duration_ms = ((end_ns - start_ns) / 1_000_000.0).round

    render json: {
      language: "ruby",
      threadId: "puma-worker-#{Process.pid}",
      startedAt: started_at,
      finishedAt: finished_at,
      durationMs: duration_ms
    }
  end

  private

  def fibonacci(n)
    return n if n <= 1
    fibonacci(n - 1) + fibonacci(n - 2)
  end
end
